src/audio/pipewire.c

```diff
@@ -0,0 +1,322 @@
+    if ((destination_memory = buf->datas[0].data) == NULL)
+        return;
+    // in the process callback, we take the samples from the ring buffer we used in the pipewire_play function and send it to pipewire for playback
+    // this will be done using the following two step algorythm:
+    // we first get the read index of the ringbuffer, which also returns the amount available in there
+    // if we have enough data in the ringbuffer, aka if the read index isn't less than requested, then we load all of it at once in pipewire
+    // however, if the remainder between the filled part of the buffer and the value of requested is greatter than 0, then we fill that part of the buffer with silence, while taking everything else, hoping we will have more samples next time
+    available_samples = spa_ringbuffer_get_read_index(&state->rb, &read_index);
+    number_of_samples = buf->datas[0].maxsize; // how many samples, in bytes,  of audio are in this chunk. Technically, requested itself could be used, but this is done instead to avoid some very rare, but important, edge cases, for example buggy hardware drivers which don't probe the card correctly, so on_process gets called very often, with a value of requested samples greatter than what pipewire could allocate, in maxsize
+#if PW_CHECK_VERSION(0, 3, 49)
+    if (b->requested > 0)
+    {
+        number_of_samples = SPA_MIN(number_of_samples, b->requested * state->stride); // because this value is given in frames, and we want samples
+    }
+#endif
+
+    // if there's something to be read from the ring buffer at this time, we do so now
+    if (available_samples > 0)
+    {
+        available_for_loading = SPA_MIN(available_samples, number_of_samples);
+    }
+    else
+    {
+        available_for_loading = 0;
+    }
+    // if there's anything else to fill and the ring buffer doesn't contain enough at this point, fill the difference with silence
+    must_load_with_silence = number_of_samples - available_for_loading;
+    if (available_for_loading > 0)
+    {
+        spa_ringbuffer_read_data(&state->rb, state->sample_buffer, SAMPLE_BUFFER_SIZE, read_index, destination_memory, available_for_loading);
+        spa_ringbuffer_read_update(&state->rb, read_index + available_for_loading);
+    }
+// if the required pipewire version doesn't match for requested to be available, we may have had plenty more time to buffer some more data, and instead we fill the whole buffer, while reading everything from the ringbuffer prematurely
+#if PW_CHECK_VERSION(0, 3, 49)
+    if (b->requested > 0)
+    {
+        if (must_load_with_silence > 0)
+        {
+            memset(destination_memory + available_for_loading, 0, must_load_with_silence);
+        }
+    }
+#endif
+    // now, we should have most of the data ready, we fill in some metadata and push the buffer object to pipewire
+    buf->datas[0].chunk->offset = 0;
+    buf->datas[0].chunk->size = number_of_samples;
```

Comment:
no, it's `available_for_loading` bytes that we have loaded in `buf`.

Reply:
I will close this thread now, if that clarifies another part of the code and there's a bug there or something, feel free to reopen

Reply:
Ah, the name of the variable `number_of_samples` mislead me, I was thinking it was always the whole buffer. But in the `(!b->requested)` case (or old pipewire version), you do want to use `available_for_loading`, since in that case you won't have filled with silence.

Reply:
I don't much care about that kind of old pipewire, if it's old enough to not have this function, it's very buggy and barely usable anyway. I may add it in the future, but we have to get actual speech out of this

Reply:
It's not just about old pipewire, but also the case when a new pipewire doesn't provide a `b->requested`

Reply:
Also, it's not that old: the previous ubuntu LTS, 22.04, has that old version of pipewire

Reply:
that's fixed now, kinda. We have no way of knowing what pw expects of us in that situation, so we make a best guess by pushing what we have, better than pushing nothing